src/kde/quick_index.cc

```diff
@@ -0,0 +1,1547 @@
+ * \param[in] domain_decomposed_
+ */
+quick_index::quick_index(const size_t dim_, const std::vector<std::array<double, 3>> &locations_,
+                         const double max_window_size_, const size_t bins_per_dimension_,
+                         const bool domain_decomposed_)
+    : dim(dim_), domain_decomposed(domain_decomposed_), coarse_bin_resolution(bins_per_dimension_),
+      max_window_size(max_window_size_), locations(locations_), n_locations(locations_.size()) {
+  Require(dim > 0);
+  Require(coarse_bin_resolution > 0);
+
+  // Build local bounding box
+  bounding_box_min = {0, 0, 0};
+  bounding_box_max = {0, 0, 0};
+  // only set initial values for working dimensions
+  for (size_t d = 0; d < dim; d++) {
+    bounding_box_min[d] = 1e20;
+    bounding_box_max[d] = -1e20;
+  }
+
+  for (auto &loc : locations) {
+    for (size_t d = 0; d < dim; d++) {
+      if (loc[d] < bounding_box_min[d])
+        bounding_box_min[d] = loc[d];
+      if (loc[d] > bounding_box_max[d])
+        bounding_box_max[d] = loc[d];
+    }
+  }
+
+  if (domain_decomposed) {
+    // Store the local bounding box and extend to maximum non-local data size
+    local_bounding_box_min = bounding_box_min;
+    local_bounding_box_max = bounding_box_max;
+    for (size_t d = 0; d < dim; d++) {
+      local_bounding_box_min[d] -= max_window_size * 0.5;
+      local_bounding_box_max[d] += max_window_size * 0.5;
+    }
+    // Global reduce to get the global min and max
+    rtt_c4::global_min(&bounding_box_min[0], 3);
+    rtt_c4::global_max(&bounding_box_max[0], 3);
+    for (size_t d = 0; d < dim; d++) {
+      local_bounding_box_min[d] = std::max(local_bounding_box_min[d], bounding_box_min[d]);
+      local_bounding_box_max[d] = std::min(local_bounding_box_max[d], bounding_box_max[d]);
+    }
+  }
+
+  // temp cast corse_bin_resolution to double for interpolation
+  const auto crd = static_cast<double>(coarse_bin_resolution);
+
+  // build up the local hash table of into global bins
+  size_t locIndex = 0;
+  for (auto &loc : locations) {
+    std::array<size_t, 3> index{0UL, 0UL, 0UL};
+    for (size_t d = 0; d < dim; d++) {
+      Check(bounding_box_min[d] < bounding_box_max[d]);
+      index[d] = static_cast<size_t>(std::floor(crd * (loc[d] - bounding_box_min[d]) /
+                                                (bounding_box_max[d] - bounding_box_min[d])));
+      index[d] = std::min(index[d], coarse_bin_resolution - 1);
+    }
+    // build up the local index hash
+    const size_t global_index = index[0] + index[1] * coarse_bin_resolution +
+                                index[2] * coarse_bin_resolution * coarse_bin_resolution;
+    coarse_index_map[global_index].push_back(locIndex);
+    locIndex++;
+  }
+
+  // Now we need to build up ghost location map data for domain decomposed mode
+  if (domain_decomposed) {
+    // temporary cast of the nodes to prevent conversion warnings
+    const auto nodes = static_cast<size_t>(rtt_c4::nodes());
```

Comment:
`rtt_c4::nranks()` returns a `uint32_t`.  Using it would allow you to remove the `static_cast`.